Add doc dir for Keptn 0.16 (#1228)
keptn/keptn.github.io: content/docs/0.16.x/bridge/oauth/index.md

---
title: OpenID Authentication
description: Enable/Disable OpenID authentication and implementation details for the identity provider
weight: 20
keywords: [0.11.x-bridge]
aliases:
  - /docs/0.11.x/operate/user_management/openid_authentication/
---

## Enable/Disable Authentication

The Keptn Bridge contains a switch to enable/disable OpenID Connect-based authentication. This switch also enables session cookies. 

> You can enable/disable OpenID via the Helm Chart values when installing Keptn setting `enabled` to `true`:

Reviewer: ```suggestion
You can enable/disable OpenID via the Helm Chart values by setting `enabled` to `true` when installing Keptn:
```

Reviewer: I just noticed that the configuration snippet that follows actually has `enabled` set to false -- isn't that misleading?

    userIdentifier: ""
```

Furthermore, additional values are required for OpenID to work:

- `baseUrl` - URL of the bridge (e.g. `http://localhost:3000` or `https://myBridgeInstallation.com`).
- `discovery` - Discovery URL of the identity provider (e.g. https://api.login.yahoo.com/.well-known/openid-configuration).
- `clientID` - Client ID, provided by the identity provider.
- `clientSecret` (optional) - Client secret, provided by the identity provider.
- `IDTokenAlg` (optional) - Algorithm that is used to verify the ID token (e.g. `ES256`). Default is `RS256`.
- `scope` (optional) - Additional scopes that should be added to the authentication flow (e.g. `profile email`), separated by space.
- `userIdentifier` (optional) - The property of the ID token that identifies the user. Default is `name` and fallback to `nickname`, `preferred_username` and `email`.

## Authentication flow with the identity provider

The following diagram shows the expected authentication flow with a custom identity provider that follows the implementation details explained below. 

{{< popup_image
link="./assets/oauth-flow.png"
caption="Authentication flow with identity provider" width="800px">}}

> Keptn Bridge will check the existence of the identity provider and also validate responses obtained from it. For example, next are requests and responses the Keptn Bridge can expect from the identity provider.

Reviewer: ```suggestion
Keptn Bridge checks the existence of the identity provider and also validates responses obtained from it. For example, the following are requests and responses the Keptn Bridge can expect from the identity provider.
```

1. HTTP 200 - Login is accepted
1. HTTP 403 - Login is unaccepted OR permission denied
1. Any other status - Considered as errors

Successful login example:

```
POST http://identity-provider:8080/token_decision
Content-Type: application/json
{
  "code": "qwert",
  "state": "123",
  "scope": "openid",
  "redirect_uri": "https://example.com/oauth/redirect,
  "client_id": "asdfxyz"
}

HTTP 200 OK
Content-Type: application/json
{
  "id_token": <id_token>,
  "access_token": <access_token>,
  "refresh_token": <refresh_token>,
  "expires_in": <expires_in>
}
```

Unsuccessful login example:

```
POST http://identity-provider:8080/token_decision
Content-Type: application/json
{
  "code": "qwery",
  "state": "123",
  "scope": "openid",
  "redirect_uri": "https://example.com/oauth/redirect,
  "client_id": "asdfxyz"
}

HTTP 403 Forbidden
Content-Type: application/json
{
  "message": "User Alex does not have permission to login to Keptn."
}
```

### jwks_uri endpoint

> It contains a set of JSON Web Keys that is used for validating the received id token.

Reviewer: ```suggestion
It contains a set of JSON Web Keys that are used for validating the received id token.
```
